jackson/src/main/java/com/lightbend/lagom/javadsl/jackson/JacksonSerializerFactory.java

```diff
@@ -116,14 +116,19 @@ public ByteString serialize(MessageEntity messageEntity) {
 
         private class JacksonDeserializer implements NegotiatedDeserializer<MessageEntity, ByteString> {
             private final ObjectReader reader;
+            private final Type type;
 
-            public JacksonDeserializer(ObjectReader reader) {
+            public JacksonDeserializer(ObjectReader reader, Type type) {
                 this.reader = reader;
+                this.type = type;
             }
 
             @Override
             public MessageEntity deserialize(ByteString bytes) {
                 try {
+                    if (bytes.isEmpty() && this.type == Optional.class) {
```

Comment:
I think it would be simpler and more consistent to always deserialize the empty body as `null`, and not require that the type is `Optional`. In this case, you wouldn't need to use the `ObjectReader` at all:

```java
if (bytes.isEmpty()) return null;
```

Reply:
In this case there will be no deserialization exception no matter what is the body type.
We agreed that if body type is not `Optional` then an error should be thrown when the body is empty.
Unless of course some other class along the way will take of handling `null`.

Reply:
As I mentioned in my previous review comment, what's implemented here is different from what I thought we agreed on (which was to handle `Option`/`Optional` request types with a top-level serializer that isn't JSON-specific), but it's more important in my mind that:

1. the Java and Scala behavior is idiomatic for each respective language,
2. behavior is consistent between languages when it doesn't conflict with 1
3. the overall behavior isn't surprising for people who haven't read the source code

(although I'd like comments from other @lagom/integrators).

Right now, the implementation of Scala serializer in this pull request always deserializes an empty byte string as `JsNull`. This means that sending an empty body (with "Content-Type: application/json") and sending a body with "null" will always do the same thing. In the case of Play JSON, by default it sticks with the Scala convention of avoiding `null` references and failing on `null` unless deserializing to `Option[T]`. This is true whether there is an empty request body or an explicit `null`.

This test fails with `JsonValidationError`:

```scala
    "convert JS null to null by default" in {
      implicit val format: Format[Dummy] = Json.format
      val dummy = JsNull.as[Dummy]
      dummy shouldBe null
    }
```

This passes:

```scala
    "convert JS null to None by default" in {
      implicit val format: Format[Dummy] = Json.format
      val dummy = JsNull.as[Option[Dummy]]
      dummy shouldBe None
    }
```

That's just the default, though, and if you really want to deserialize to `null` then you can define your JSON reader to do so. This also passes:

```scala
    "convert JS null to null with custom Format" in {
      implicit val format: Format[Dummy] = new Format[Dummy] {
        override def writes(o: Dummy): JsValue = o match {
          case Dummy(prop) => JsObject(Seq(("prop", Json.toJson(prop))))
          case null        => JsNull
        }

        override def reads(json: JsValue): JsResult[Dummy] = json match {
          case o: JsObject => JsSuccess(Dummy((o \ "prop").asOpt[String]))
          case JsNull      => JsSuccess(null)
        }
      }
      val dummy = JsNull.as[Dummy]
      dummy shouldBe null
    }
```

Special casing `Optional` in the Java version isn't consistent. Sending an empty body behaves differently than sending a body containing "null". Jackson, by default, will deserialize "null" from JSON to a `null` Java reference, which is more conventional for Java.

This passes:

```java
    @Test
    public void testDeserializeNullToDummy() {
        StrictMessageSerializer<Dummy> serializer = factory.messageSerializerFor(Dummy.class);
        Dummy dummy = serializer.deserializer(new MessageProtocol().withContentType("application/json"))
            .deserialize(ByteString.fromString("null"));
        assertNull(dummy);
    }
```

So does this:

```java

    @Test
    public void testDeserializeNullToOptionalDummy() {
        StrictMessageSerializer<Optional<Dummy>> serializer = factory.messageSerializerFor(Optional.class);
        Optional<Dummy> dummy = serializer.deserializer(new MessageProtocol().withContentType("application/json"))
            .deserialize(ByteString.fromString("null"));
        assertEquals(Optional.empty(), dummy);
    }
```

What I'm suggesting is that the empty body and a body with "null" should always behave the same way as each other, regardless of language, even though there are other differences in the way `null` is handled by default between the languages.

Reply:
> As I mentioned in my previous review comment, what's implemented here is different from what I thought we agreed on (which was to handle Option/Optional request types with a top-level serializer that isn't JSON-specific)

@TimMoore, could you please elaborate here what was your original thought?

Defining a serializer that inherits from `StrictMessageSerializer` and then `JsValueMessageSerializer` can inherit from the new one instead of strict?

I know that this PR might require more guidance than we both thought :face_with_head_bandage:

Reply:
Sorry, bear with me because I haven't really fully thought it out.

My original thought was that there would be a generic optional serializer like:

```scala
implicit def optionMessageSerializer[A](implicit valueSerializer: MessageSerializer[A, ByteString]): StrictMessageSerializer[Option[A]] = new StrictMessageSerializer[Option[A]] {
  // ...
  private class OptionalDeserializer(valueDeserializer: NegotiatedDeserializer[A, ByteString]) extends NegotiatedDeserializer[Option[A], ByteString] {
    override def deserialize(wire: ByteString): Option[A] = {
      if (wire.isEmpty) None
      else Some(valueDeserializer.deserialize(wire))
    }
  }
  // ...
}
```

This wouldn't be specific to JSON, it would just always deserialize an empty body as `None` and delegate a non-empty body to another deserializer.

It would have to work a bit differently in the Java API because it doesn't have implicits. I haven't done a really detailed investigation on how to make it work there.

Reply:
In this case I will have to explicitly define that I would like to use this serializer in the service call description which is of course an option but is it what we are aiming for?